src/tasks/PageIterator.ts

```diff
@@ -0,0 +1,341 @@
+ */
+export interface PageCollection<T> {
+  value: T[];
+  odataNextLink?: string;
+  odataDeltaLink?: string;
+  [Key: string]: any;
+}
+
+/**
+ * Signature representing callback for page iterator
+ * @property {Function} callback - The callback function which should return boolean to continue the continue/stop the iteration.
+ */
+export type PageIteratorCallback<T> = (data: T) => boolean;
+
+/**
+ * Signature to define the request options to be sent during request.
+ * The values of the GraphRequestOptions properties are passed to the Graph Request object.
+ * @property {Headers} headers - the header options for the request
+ * @property {RequestOption[]} options - The middleware options for the request
+ */
+export interface PagingRequestOptions {
+  headers?: Headers;
+  requestOption?: RequestOption[];
+}
+
+/**
+ * @enum
+ * Enum representing the state of the iterator
+ */
+export enum PagingState {
+  NotStarted,
+  Paused,
+  IntrapageIteration,
+  InterpageIteration,
+  Delta,
+  Complete,
+}
+
+/**
+ * @class
+ * Class for PageIterator
+ */
+export class PageIterator<T extends Parsable, C extends Parsable> {
+  /**
+   * @private
+   * Member holding the GraphClient instance
+   */
+  private readonly requestAdapter: RequestAdapter;
+
+  /**
+   * @private
+   * Member holding the current page
+   */
+  private currentPage?: PageCollection<T>;
+
+  /**
+   * @private
+   * Member holding a complete/incomplete status of an iterator
+   */
+  private complete: boolean;
+
+  /**
+   * @private
+   * Member holding the current position on the collection
+   */
+  private cursor: number;
+
+  /**
+   * @private
+   * Member holding the headers that can be added to the request
+   */
+  private readonly headers: Headers;
+
+  /**
+   * @private
+   * Member holding the factory to create the parsable object
+   */
+  private readonly parsableFactory: ParsableFactory<C>;
+
+  /**
+   * @private
+   * Member holding the error mappings
+   */
+  private readonly errorMappings: ErrorMappings;
+
+  /*
+   * @private
+   * Member holding the callback for iteration
+   */
+  private readonly callback: PageIteratorCallback<T>;
+
+  /**
+   * @private
+   * Member holding the state of the iterator
+   */
+  private pagingState: PagingState;
+
+  /**
+   * @public
+   * @constructor
+   * Creates new instance for PageIterator
+   * @returns An instance of a PageIterator
+   * @param requestAdapter - The request adapter
+   * @param pageResult - The page collection result
+   * @param callback - The callback function to be called on each item
+   * @param errorMappings - The error mappings
+   * @param parsableFactory - The factory to create the parsable object
+   * @param options - The request options to configure the request
+   */
+  public constructor(
+    requestAdapter: RequestAdapter,
+    pageResult: C,
```

Comment:
can't we express C as just "something<T>" instead of requiring both type params?

Reply:
C is a collection of paged items. Its supposed to help differentiate the expected parsable factory

Reply:
I suppose we can also leverage an interface similar to the `BaseDeltaFunctionResponse` available in the generated models?